### ISSUE TITLE
[NICE FIRST CONTRIB] Streams operators docs - examples & diagrams

### ISSUE THREAD

**Issue by ktoso** [akka/akka#24933]
In the resolution PR to https://github.com/akka/akka/issues/24572 we are moving towards "one page per operator", e.g. Flow.map gets it's own documentation page as do all other FileIO and all other Flow/Source/Sink operators.

We want to have an example for each of them, as well as an "marble" diagram (or similar) for each of them; we will require the help of the community to make the initial addition happen since there are quire a few operators.

We propose using draw.io for all the diagrams; We should provide a common set of "images" or at least an initially consistent style for them. This is one example of working on a map function image:
<img width="802" alt="map_xml_-_draw_io" src="https://user-images.githubusercontent.com/120979/39027228-e6a04b02-4448-11e8-81c6-a71e5fd356be.png">

TODO: Propose a specific workflow how to add such diagrams. The resulting XML file, such as here: https://github.com/ktoso/akkamazing/blob/master/src/Flow-or-Source/map.xml should be stored right next to the actual png / svg that we generate.

TODO LIST ALL OPERATORS TO BE DOCUMENTED HERE

**Comment by hepin1989** [comment]
Yeah，That would be very great!

**Comment by ktoso** [comment]
please contribute ;)

**Comment by richardimaoka** [comment]
I think diagrams should include Source/Flow/Sink shapes as in [Basics of composition and modularity](https://doc.akka.io/docs/akka/current/stream/stream-composition.html#basics-of-composition-and-modularity), because some operators are not simple input/output but connect different graph stages (e.g. `alsoTo` below, and [link to the draw.io file](https://drive.google.com/open?id=1Xw5ocJkYjj11WlIL51al8E9z6hGcP6A9))
![2018-04-23_02h34_57](https://user-images.githubusercontent.com/7414320/39097998-1a2cb610-469f-11e8-900a-70da8cc094de.png)

For the TODOs, there will be a lot to discuss, so if we should move to gitter or whatever channel until more concrete plans come out, please let us know. Otherwise, let's continue the discussion here.

(e.g.) 
where to store the draw.io files, not only the resulting .xml
should draw.io files be versioned?
first set of operators to tackle, to see what the consistent style should be
how we incorporate the backpressure behavior into diagrams, or do we not do that to simplify the diagrams?
etc

Happy to help here 😃

**Comment by ktoso** [comment]
> where to store the draw.io files, not only the resulting .xml

Right next to the image in the akka/akka repo indeed. They should be part of the PR, the xml and the image

> should draw.io files be versioned?

I believe the xml files are the "full draw.io info" such that one can "open from github" and edit them there.

> first set of operators to tackle, to see what the consistent style should be

The simple linear ones, filter, map etc... Let's find a style and method and go from there

> how we incorporate the backpressure behavior into diagrams, or do we not do that to simplify the diagrams?

Let's not include for those where not needed; plenty operators are rather simple "just works"™, but the ones like conflate / expand indeed should have some way of marking it. We need to make that way ™ consistent

**Comment by ktoso** [comment]
Very excited about your help here, thanks a lot!

**Comment by raboof** [comment]
![2018-04-20 02 16 48](https://user-images.githubusercontent.com/131856/39117240-d071231a-46e6-11e8-9305-53aff1ae08f7.png)
I'm not sure I'm comfortable with this though... but perhaps we can work around needing this permission?

**Comment by raboof** [comment]
One option might be something like the 'rx marbles', but also showing backpressure. For example (ascii-art):
```
-1-2--3--->
|--~~-~~~---|
| broadcast |
|------------|
-1~~2---3->
-1--2---3->
-1--2~~~3->
```

**Comment by patriknw** [comment]
@raboof It should be possible to use draw.io without giving it github access. I think files can be imported/exported.

**Comment by ktoso** [comment]
I just figured out that we should call the "backpressures when" etc sections as "Reactive Streams Semantics", it will help establish Akka Streams == Reactive Streams.

<img width="665" alt="actorrefwithack_ _akka_documentation" src="https://user-images.githubusercontent.com/120979/39706098-4501a79e-524b-11e8-8321-482530206c78.png">

So such heading on all operators pages before the "... when ..." section

**Comment by johanandren** [comment]
A few operators/set of operators that are completely missing in the overview right now (some of them has documentation in other places in the streams section of the docs though):

- [ ] `akka.stream.{scaladsl,javadsl}.Compression`
- [ ] `akka.stream.{scaladsl,javadsl}.CoupledTerminationFlow`
- [ ] `akka.stream.{scaladsl,javadsl}.Framing`
- [ ] `akka.stream.{scaladsl,javadsl}.JsonFraming`
- [ ] `akka.stream.{scaladsl,javadsl}.MergeHub`
- [ ] `akka.stream.{scaladsl,javadsl}.PartitionHub`
- [ ] `akka.stream.{scaladsl,javadsl}.StreamRefs`
- [ ] `akka.stream.{scaladsl,javadsl}.Tcp`

**Comment by raboof** [comment]
Also the methods listed in https://github.com/akka/akka/blob/master/project/StreamOperatorsIndexGenerator.scala#L43 are missing and should be documented (or moved to the 'ignore' list if we decide they should not be in there).

**Comment by idarlington** [comment]
Hi I will be working on adding examples to `Source.fromFuture`

**Comment by jkransen** [comment]
I'll see how far I can get with `Source/Flow.sliding`

**Comment by brightonk** [comment]
Hi we are working on this akka.stream.{scaladsl,javadsl}.Compression

**Comment by iyogi** [comment]
Adding Source.single

**Comment by ktoso** [comment]
Thanks a lot for everyone's help here so far!

**Comment by jroper** [comment]
I've been working on marble diagrams for MicroProfile Reactive Streams Operators. One goal I have is to not use a visual diagram editor, since lets say you create 100 diagrams, and then you decide that you need to tweak the way they look for the 101st. What are you going to do, go back and hand edit 100 diagrams with your mouse is your visual editor? Or live with inconsistent diagrams? Neither. Diagrams should be specified in code, and generated. Then you can tweak all diagrams at once. And it turns out that this is very easy to do, especially if you use SVG manipulated by JavaScript, since after editing either the DSL or the generator, all you need to do is hit refresh in your browser to see the results.

Here's the result of about 4 hours work, the DSL:

```javascript
{
  map: {
    stages: [
      ins(n(1),  n(5),  n(3),  term ),
      op(".map(i -> i * 10)"),
      out(n(10), n(50), n(30), term )
    ]
  },
  filter: {
    stages: [
      ins(n(1), n(7), n(5), n(2), term ),
      op(".filter(i -> i < 3)"),
      out(n(1), none, none, n(2), term )
    ]
  },
  flatMap: {
    stages: [
      ins(n(1), none, none, none, n(2), none, none, none, term ),
      sub(none, n(4), n(5), term                               ),
      sub(none, none, none, none, none, n(3), n(6), term       ),
      op(".flatMap(i -> stream[i])"),
      out(none, n(4), n(5), none, none, n(3), n(6), none, term )
    ]
  }
}
```

And the generated diagrams:

![marbles](https://user-images.githubusercontent.com/105833/43992011-3b581d20-9dbb-11e8-8450-57481b6a3dac.png)

The code is in here:

https://github.com/jroper/microprofile-reactive-streams/tree/marble-diagrams

My plan is to create a runner that uses phantomjs to render the diagrams, and then export them, so they can be referenced from documentation.

**Comment by ktoso** [comment]
This is pretty good indeed! If we could embed it simply in code blocks in docs pages that would be awesome :) Are you planning on working on these somehow now?

**Comment by jroper** [comment]
It's definitely possible to embed the JS directly in the docs, though it would mean also pulling in the supporting libraries (two, the marbles dsl and svg.js) into all the documentation. Also not sure if there's any differences in browser support for the svg features used, the high level library I use (svg.js) does do some interrogation of the browser SVG APIs to calculate things like leading, and I think there may be some differences here. If you export to PNG, you are guaranteed to have the same image everywhere.

Another advantage of having the JS outside the docs is you can reference the resulting images from multiple places, eg from the guide, from scaladocs, javadocs, etc.

Have a look at https://github.com/eclipse/microprofile-reactive-streams/pull/90 for a complete picture.

**Comment by raboof** [comment]
Very cool, that looks really promising @jroper!

For some of the more exotic operators we might still have to post-process the generated images, but that's still a lot easier to maintain than doing everything by hand.

It would be kind of neat to also be able to express backpressure in the diagram, which seems possible by extending the dsl a bit, but let's take it one step at a time.

I think we discuss just the diagrams in this issue, I moved the other topics to separate issues:

* https://github.com/akka/akka/issues/25466 (new sets of operators)
* https://github.com/akka/akka/issues/25467 (new operators)
* https://github.com/akka/akka/issues/25468 (new examples)

Generating (SVG or PNG) as part of the build would be neat indeed.

**Comment by jroper** [comment]
For backpressure, I'd say it should be included in a per diagram basis, eg it makes a lot of sense to visualise backpressure in `lazySource` or `batch`, but for `map` or `takeWhile` it's going to make the diagrams needlessly complex. It wouldn't be hard to add something that described the backpressure, eg arrows going the other way representing a pull.

For post processing - I was thinking of offering a way to plug in custom rendering code to annotate images after they've been generated. It's not very hard to write code to generate SVG, and the code might be able to interrogate the graph to find the positions of certain marbles etc to render extra lines or highlight certain areas of the diagram, etc.

**Comment by raboof** [comment]
Yes, both 'manual' and automatic post-processing are interesting - after all while we don't want to rework 100 diagrams when we want to tweak the style, but we're not going to tweak the style all the time either, so there's a trade-off there ;).

**Comment by patriknw** [comment]
I like this approach. Makes the maintenance burden less scary, and shouldn't be a good way to get consistent images in contributions.

**Comment by jroper** [comment]
The main risk (and perhaps we should let it incubate for a while in this MicroProfile project to see if that risk materializes) is that we end up with a pile of JS that only one person (at the moment, that's me) can maintain. While the DSL is simple, it's very dependent on its implementation details which only I know because I wrote the code.

What I'd probably suggest doing is not taking what I've done in MicroProfile, but rather rewriting in ScalaJS. ScalaJS will give you a far, far nicer and easier to write DSL, and be much easier for the Akka team to maintain. Meanwhile, I don't think I can use ScalaJS here in MicroProfile, because it has to be maintained by people who have an aversion to anything Scala.

A third alternative would be to create a new svg-marbles project that uses ScalaJS that both can use, but the overhead of doing that might be more than any of us are willing to accept - I only did this because after a quick spike I saw that I could hack something half decent up in less than a week. Making a proper project I think would add another week or so when taking into account the friction introduced by having the code outside of my project.

**Comment by jroper** [comment]
Here's the finished product for MicroProfile:

https://jroper.github.io/microprofile-reactive-streams/apidocs/?org/eclipse/microprofile/reactive/streams/PublisherBuilder.html
https://jroper.github.io/microprofile-reactive-streams/apidocs/?org/eclipse/microprofile/reactive/streams/ReactiveStreams.html

The example diagram displayed at the top, with all the labels, is an example of a diagram generated that has had arbitrary post processing applied. In that particular case, the post processing should be reasonably tolerant to tweaks made to the diagram layout and sizing, it's able to find the location of elements that it wants to pin its labels to, and then position labels relative to that. The lines to the labels are rendered automatically.

**Comment by jrudolph** [comment]
@jroper is there any reason why this needs to be implemented in javascript? Would there be any disadvantage of "just generating" a svg file from a description / DSL?

**Comment by jroper** [comment]
It's not just using JavaScript, it's using a browser. For example, how do you calculate the position and size of a box that is meant to surround some centre aligned text? You need to know how big the bounding box for the text will be, and what its position will be. I'm not aware of any libraries in anything other than JavaScript that will give you that, with JavaScript, it's easy, you ask the browser to render it for you, then ask the browser for the extents of the text. And then there's JavaScript libraries, like the one I've used, that make that a one liner. Implementing that without a browser is going to be far more effort than it's worth.

**Comment by jroper** [comment]
Also, it's really nice to be able to just open the code in a browser, and using the developer tools, adjust sizes and parameters of the SVG directly in the DOM, and see the results as you type, makes all the tedious work of styling images - and that's really what the thing that we want to solve here is - very simple.

**Comment by tg44** [comment]
maybe: https://github.com/ReactiveX/rxjs/pull/757
They not just generates marbles from the input, but they can run it as a [testcase](https://github.com/ReactiveX/rxjs/pull/757/files#diff-160d21d14675aa5883ba4500373a75beR173) :D

BTW I don't think that just bcs of the styling we need the comfort(?) of the browser.

(But the basic idea is really good!)

**Comment by tg44** [comment]
Theorycrafting:
As I understands the `Materializer` is a magical creature which handles the actual creation of the graph and somewhat responsible for the proper dataflow between the inlets and outlets (or if I'm completely wrong we still can achive the most of the idea with an overriden builder which injects a logger flow between all inlet and outlet connection).
So if we have a materializer which can only materialize one graph in its life, but log every inner dataflow to an inner state from that graph run, we gain an ability to replay a concrete graph run.
If we can get out the graph edges, and the name attributes too; we can write a tool which can visualize the whole process as an animation or a single marble picture too. We can visualize pulls and pushes too. With some tweeks we can show only one stage from a bigger graph so we can visualize only a `map` or `filter` stage, but we can visualize whole graphs for deadlock searching/demonstrating (hello cycles) reasons too. If we add more attributes to the concrete flows we can tell the visualizer the parameters of the stage (so we can write `map(_ * 2)` inside a box on a picture).

So in theory we can write:
 - a materializer/builder which can persist the graph layout and one run of a graph to an inner state
 - a visualizer which can create pictures (bunch of jpeg, or gif) from the above described state
 - testlike codes to show the users how each stage emits/pulls/completes (we can even use some prewritten tests)
 - some code which regenerate the viewable thing to the doc when the doc generated

This will work for:
 - about the half of the sources (timedriven, actorref, fromfuture will not work this way)
 - about half of the sinks (head, last, seq; is easy, fold will be problematic bcs we don't see inner states)
 - most of the simple operators (async, future; problematic)
 - most of the fan-in and fan-out

This will problematic for:
 - time aware things (we don't see the inner states so kinda blackbox game, but with some effort we can show the buffer overflow strategies)
 - async things or everything which can create future/source from an element
 - time driven things (maybe solvable but not as trivial)
 - nesting and flattening maybe works out of the box, maybe not, no idea without the proper knowledge of underlying code
 - anything with io/logging/promise/actor

**Comment by johanandren** [comment]
@tg44 I don't think that it would work out to do an automagically generated diagram, it will need to be designed by a person that understands the purpose of the operator does and what is important to convey to the reader about it.

**Comment by tg44** [comment]
We already have a good dsl to manually trigger events on a stream. If we use the TestSource.probe for example we can manually write the "person designed" snippets which will run on the special materializer and converted to a picture/animation. 
So I would not automate the actual snippets. I would rather use a prewritten dsl/ecosystem (testkit, streams) to generate the output picture/animation. I think the possibility to generate diagram from any graph not means that we don't need to carefully construct the examples in the docs.

But I would rather use a syntax like
```
Source(List(1,2,3)).map(_ * 2).addAttributes(Attributes.describe("_ * 2")).runWith(Sink.ignore)(mapExampleMat)
val img = MarbleGenerator(mapExampleMat.states).createJpg
```
Than sth like the RxJs do.

Soo its interesting enough, I will do an example project to see if I can do it or not, or maybe the code will describe more what I meant.

**Comment by johanandren** [comment]
Having used the dot/graphviz output we already have in place for the stream testkit (when `assertAllStages` fails) I'd say that is very close to no value for a new-comer who is trying to learn streams (very useful for someone writing stages and trying to figure out what is wrong though).
